### ISSUE TITLE
Improve error message when using `TypeVar` in functions

### ISSUE THREAD

**Issue by Dreamsorcerer** [python/mypy#12012]
We're trying to add static typing support to the `Application` object in aiohttp. Currently the latest proposal is to define a custom key which can be used something like:
```
k = AppKey[int]("k")
app[k] = "foo"
```

The method definition looks like:
```
def __setitem__(self, key: AppKey[_T], value: _T) -> None:
```

The problem is that example produces a rather uninformative error:
```
Cannot infer type argument 1 of "__setitem__" of "Application"  [misc]
```

I'd really like to see something along the lines of `"str" is not compatible with "int"`.

Related PR: https://github.com/aio-libs/aiohttp/pull/6498

**Comment by erictraut** [comment]
@Dreamsorcerer, could you provide a minimal self-contained code sample that demonstrates the problem?

**Comment by Dreamsorcerer** [comment]
Yep, the AppKey thing is available in aiohttp's 3.9 branch. But, this seems to work as a reproducer:

```
from typing import Generic, TypeVar

T = TypeVar("T")

class AppKey(Generic[T]):
    ...


class Application:
    def __setitem__(self, key: AppKey[T], value: T) -> None:
        ...


a = Application()
k = AppKey[int]()

a[k] = "foo"  # test.py:17: error: Cannot infer type argument 1 of "__setitem__" of "Application"  [misc]
```

I would expect the error to atleast mention `int` or something more useful to the user. (`__setitem__ expected int, but received str` or something).

**Comment by Olegt0rr** [comment]
Reproducible example: https://github.com/aio-libs/aiohttp-sse/actions/runs/7696889151/job/20972675636?pr=441
